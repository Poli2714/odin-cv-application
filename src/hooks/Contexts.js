import { createContext } from 'react';

const IndexContext = createContext({
  activeIndex: null,
  setActiveIndex: () => {},
});

const KeyContext = createContext({
  activeKey: {
    eduActiveKey: null,
    expActiveKey: null,
  },
  setActiveKey: () => {},
});

const EduErrMsgContext = createContext({
  showEduErrMessage: false,
  setShowEduErrMessage: () => {},
});

const ExpErrMsgContext = createContext({
  showExpErrMessage: false,
  setShowExpErrMessage: () => {},
});

const PersonalInfoContext = createContext({
  personalInfo: {
    profilePicture: '',
    fullName: '',
    currentJobTitle: '',
    email: '',
    mobile: '',
    city: '',
  },
  setPersonalInfo: () => {},
});

const SummaryContext = createContext({
  summary: '',
  setSummary: () => {},
});

const ExperienceContext = createContext({
  experiences: {
    id: {
      jobTitle: '',
      employer: '',
      jobStartDate: '',
      jobEndDate: '',
      jobDescription: '',
    },
  },
  setExperiences: () => {},
});

const EducationContext = createContext({
  educations: {
    id: {
      degree: '',
      university: '',
      eduStartDate: '',
      eduEndDate: '',
      eduInfo: '',
    },
  },
  setEducations: () => {},
});

export {
  EducationContext,
  EduErrMsgContext,
  ExperienceContext,
  ExpErrMsgContext,
  IndexContext,
  KeyContext,
  PersonalInfoContext,
  SummaryContext,
};
