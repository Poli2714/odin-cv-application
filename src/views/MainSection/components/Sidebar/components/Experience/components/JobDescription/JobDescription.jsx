import { useContext } from 'react';
import PropTypes from 'prop-types';

import { CharacterLimit } from 'src/components/ui';
import { InputWrapper } from 'src/components/layout';
import { Label, TextArea } from 'src/components/forms';
import { ExperienceContext } from 'src/hooks/Contexts';

const limit = 200;
const rows = 10;

function JobDescription({ id }) {
  const { experiences, setExperiences } = useContext(ExperienceContext);

  const handleChangeJobDescription = (e, key) =>
    setExperiences({
      ...experiences,
      [key]: { ...experiences[key], jobDescription: e.target.value },
    });

  return (
    <InputWrapper>
      <Label labelName="Job Description" labelFor="jobDescription" />
      <CharacterLimit limit={limit} value={experiences[id].jobDescription}>
        <TextArea
          name="jobDescription"
          placeholder="Write your job description"
          rows={rows}
          maxLength={limit}
          value={experiences[id].jobDescription}
          onChange={e => handleChangeJobDescription(e, id)}
        />
      </CharacterLimit>
    </InputWrapper>
  );
}

JobDescription.propTypes = {
  id: PropTypes.string,
};

export default JobDescription;
