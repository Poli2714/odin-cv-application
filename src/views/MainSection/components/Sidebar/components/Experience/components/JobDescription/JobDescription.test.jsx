import { expect, test, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useState } from 'react';

import JobDescription from './JobDescription';
import { ExperienceContext } from 'src/hooks/Contexts';

vi.mock('src/components/ui', () => ({
  CharacterLimit: vi.fn(({ children, value }) => (
    <div>
      <span data-testid="limit">{value.length} / 200</span>
      {children}
    </div>
  )),
}));

vi.mock('src/components/layout', () => ({
  InputWrapper: vi.fn(({ children }) => (
    <div data-testid="input-wrapper">{children}</div>
  )),
}));

vi.mock('src/components/forms', () => ({
  Label: vi.fn(() => <label htmlFor="jobDescription">Mock label</label>),
  TextArea: vi.fn(({ onChange, value }) => (
    <textarea
      id="jobDescription"
      rows={10}
      maxLength={200}
      onChange={onChange}
      value={value}
    ></textarea>
  )),
}));

function MockParentComponent({ id }) {
  const [experiences, setExperiences] = useState({
    testId: { jobDescription: '' },
    testId2: { jobDescription: 'Instagram' },
  });

  return (
    <ExperienceContext.Provider value={{ experiences, setExperiences }}>
      <JobDescription id={id} />
    </ExperienceContext.Provider>
  );
}

test('renders JobDescription', () => {
  render(<MockParentComponent id="testId" />);

  expect(screen.getByTestId('input-wrapper')).toMatchSnapshot();
});

test('renders initial jobDescription value correctly', () => {
  render(<MockParentComponent id="testId2" />);

  expect(screen.getByLabelText('Mock label')).toHaveValue('Instagram');
  expect(screen.getByTestId('limit')).toHaveTextContent('9 / 200');
});

test('updates jobDescription value as a user types', async () => {
  const user = userEvent.setup();

  render(<MockParentComponent id="testId" />);
  const textarea = screen.getByLabelText('Mock label');
  const limitSpan = screen.getByTestId('limit');

  expect(textarea).toHaveValue('');
  expect(limitSpan).toHaveTextContent('0 / 200');
  await user.type(textarea, 'Cool');
  expect(textarea).toHaveValue('Cool');
  expect(limitSpan).toHaveTextContent('4 / 200');
  await user.clear(textarea);
  await user.type(textarea, 'Hello');
  expect(textarea).toHaveValue('Hello');
  expect(limitSpan).toHaveTextContent('5 / 200');
});

test('renders jobDescription value with correct number of characters when user reaches the limit', async () => {
  const user = userEvent.setup();

  render(<MockParentComponent id="testId" />);
  const textarea = screen.getByLabelText('Mock label');
  const limitSpan = screen.getByTestId('limit');

  await user.type(
    textarea,
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit. In sed posuere nisl, in dignissim metus. Fusce maximus commodo iaculis. Sed placerat fringilla risus, sit amet malesuada lectus bibendum at. Pellentesque ac ex eu.',
  );

  expect(textarea.value).toHaveLength(200);
  expect(limitSpan).toHaveTextContent('200 / 200');
});
