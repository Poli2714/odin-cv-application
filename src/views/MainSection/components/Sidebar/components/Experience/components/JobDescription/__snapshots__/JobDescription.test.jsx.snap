// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders JobDescription 1`] = `
<div
  data-testid="input-wrapper"
>
  <label
    for="jobDescription"
  >
    Mock label
  </label>
  <div>
    <span
      data-testid="limit"
    >
      0
       / 200
    </span>
    <textarea
      id="jobDescription"
      maxlength="200"
      rows="10"
    />
  </div>
</div>
`;
